File: chroma-league/src/test/java/com/bonepl/chromaleague/hud/parts/resource/RumbleHeatBarTest.java
package com.bonepl.chromaleague.hud.parts.resource;

import com.bonepl.chromaleague.IntSteps;
import com.bonepl.chromaleague.hud.AnimationTester;
import org.junit.jupiter.api.Test;

class RumbleHeatBarTest {

    @Test
    void testRumbleStandardHeat() {
        final IntSteps heatSteps = new IntSteps(0, 100, 5);
        final IntSteps cooldownSteps = new IntSteps(100, 0, 2);

        final RumbleHeatBar rumbleHeatBar = new RumbleHeatBar();
        new AnimationTester()
                .withBeforeIterationAction(i -> {
                    if (i < 20) {
                        ManaBarTest.mockResource(heatSteps.nextInt(), 100);
                    } else {
                        ManaBarTest.mockResource(cooldownSteps.nextInt(), 100);
                    }
                })
                .withSleepTime(300)
                .testAnimation(rumbleHeatBar, 67);
    }

    @Test
    void testRumbleOverheatHeat() {
        final IntSteps heatSteps = new IntSteps(0, 100, 5);

        final RumbleHeatBar rumbleHeatBar = new RumbleHeatBar();
        new AnimationTester()
                .withBeforeIterationAction(i -> {
                    ManaBarTest.mockResource(heatSteps.nextInt(), 100);
                })
                .withSleepTime(300)
                .testAnimation(rumbleHeatBar, 40);
    }
}

File: test/hud/parts/resource/RumbleHeatBar.test.ts
import {describe, test} from 'vitest';
import {IntSteps} from '../../../IntSteps.js';
import {AnimationTester} from '../../../AnimationTester.js';
import {RumbleHeatBar} from '../../../../src/hud/parts/resource/RumbleHeatBar.js';
import {mockResource, setupResourceTest} from './AbstractResourceTest.js';

/** Integration test for Rumble heat resource bar on actual hardware. */
describe('RumbleHeat Integration', () => {
    setupResourceTest();

    test('testRumbleStandardHeat', async () => {
        const heatSteps = new IntSteps(0, 100, 5);
        const cooldownSteps = new IntSteps(100, 0, 2);

        const rumbleHeatBar = new RumbleHeatBar();
        await new AnimationTester()
            .withBeforeIterationAction(i => {
                if (i < 20) {
                    mockResource(heatSteps.nextInt(), 100);
                } else {
                    mockResource(cooldownSteps.nextInt(), 100);
                }
            })
            .withSleepTime(300)
            .testAnimation(rumbleHeatBar, 67);
    }, 30000);

    test('testRumbleOverheatHeat', async () => {
        const heatSteps = new IntSteps(0, 100, 5);

        const rumbleHeatBar = new RumbleHeatBar();
        await new AnimationTester()
            .withBeforeIterationAction(() => mockResource(heatSteps.nextInt(), 100))
            .withSleepTime(300)
            .testAnimation(rumbleHeatBar, 40);
    }, 30000);
});
